core/unix/coredump.c

```diff
@@ -0,0 +1,360 @@
+    ELF_OFF file_offset = sizeof(ELF_HEADER_TYPE) +
+        1 /*program_header_count*/ * sizeof(ELF_PROGRAM_HEADER_TYPE);
+    for (int section_index = 0; section_index < section_count - 1; ++section_index) {
+        ELF_WORD flags = SHF_ALLOC | SHF_MERGE;
+        if (section_header_info[section_index].prot & PROT_WRITE) {
+            flags |= SHF_WRITE;
+        }
+        if (!write_section_header(
+                elf_file, section_header_info[section_index].name_offset, SHT_PROGBITS,
+                flags, (ELF_ADDR)section_header_info[section_index].vm_start, file_offset,
+                section_header_info[section_index].vm_end -
+                    section_header_info[section_index].vm_start,
+                /*link=*/0,
+                /*info=*/0, /*alignment=*/sizeof(ELF_WORD),
+                /*entry_size=*/0)) {
+            os_close(elf_file);
+            return false;
+        }
+        file_offset += section_header_info[section_index].vm_end -
+            section_header_info[section_index].vm_start;
+    }
+    // Write the section names section.
+    if (!write_section_header(
+            elf_file, string_table_offset - strlen(SECTION_HEADER_TABLE), SHT_STRTAB,
+            /*flags=*/0, /*virtual_address=*/0, file_offset,
+            /*section_size=*/string_table_offset, /*link=*/0,
+            /*info=*/0, /*alignment=*/1,
+            /*entry_size=*/0)) {
+        os_close(elf_file);
+        return false;
+    }
+
+    os_close(elf_file);
+    strhash_hash_destroy(GLOBAL_DCONTEXT, string_htable);
+    return true;
+}
+
+/*
+ * Returns true if a core dump file is written, false otherwise.
+ */
+bool
+os_dump_core_live(void)
+{
+    static thread_id_t current_dumping_thread_id VAR_IN_SECTION(NEVER_PROTECTED_SECTION) =
+        0;
+    thread_id_t current_id = d_r_get_thread_id();
+#ifdef DEADLOCK_AVOIDANCE
+    dcontext_t *dcontext = get_thread_private_dcontext();
+    thread_locks_t *old_thread_owned_locks = NULL;
+#endif
+
+    if (current_id == current_dumping_thread_id) {
+        return false; /* avoid infinite loop */
+    }
+
+#ifdef DEADLOCK_AVOIDANCE
+    /* First turn off deadlock avoidance for this thread (needed for live dump
+     * to try to grab all_threads and thread_initexit locks).
+     */
+    if (dcontext != NULL) {
+        old_thread_owned_locks = dcontext->thread_owned_locks;
+        dcontext->thread_owned_locks = NULL;
+    }
+#endif
+    /* Only allow one thread to dumpcore at a time, also protects static
+     * buffers and current_dumping_thread_id.
+     */
+    d_r_mutex_lock(&dump_core_lock);
+    current_dumping_thread_id = current_id;
+    const bool ret = os_dump_core_internal();
+    current_dumping_thread_id = 0;
+    d_r_mutex_unlock(&dump_core_lock);
+
+#ifdef DEADLOCK_AVOIDANCE
+    /* Restore deadlock avoidance for this thread. */
+    if (dcontext != NULL) {
+        dcontext->thread_owned_locks = old_thread_owned_locks;
+    }
+#endif
+    return ret;
+}
```

Comment:
It would be nice to add an automated test of this.

Reply:
I agree. How about adding an automated test in a followup PR?

Reply:
Generally it's best to have tests included in the same change as the feature.  Could we at least have a sanity test that exercises the feature even if it doesn't validate every piece of the resulting file.

Reply:
I've added a new test, memory_dump_test, to cover dr_create_memory_dump for X64 Linux.

Reply:
Just to clarify I'm only requesting adding a simple test that calls dr_create_memory_dump(); TODOs in the test for actually checking contents seems fine to be added later.  Just want to at least call the API routine in a test.  (This is also tech debt as the original feature should have added such a test; but best to have at least a test exercising a new feature in the new feature PR.)

Reply:
I've added a TODO in suite/tests/CMakeLists.txt.

Reply:
I've moved the TODO to the test file.